### ISSUE TITLE
Conditional types break with property chaining

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#32735]
**TypeScript Version:**  3.5.1

**Search Terms:**

property chaining, generic, object type param, conditional type

**Code**

Modified version of @fatcerberus ' attempt to break TS.

```ts
type Droste<T extends {x:number|string}> = {
    value: T,
    /**
     * Should alternate between string and number
     */
    droste: Droste<{x:(T["x"] extends number ? string : number)}>
};

declare const droste: Droste<{x:number}>;

//number
const _0 = droste.droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste;
const _1 = _0.droste; //string
const _2 = _1.droste; //number
const _3 = _2.droste; //string

const _4 = _3.droste; //Expected number, actual string, ???
const _5 = _4.droste; //string
const _6 = _5.droste; //string
const _7 = _6.droste; //string
const _8 = _7.droste; //string
const _9 = _8.droste; //string
//string forever and ever. Where is `number`?
```

**Expected behavior:**

Each access to `.droste` should alternate between `string` and `number`.
Or give a max instantiation depth/count error.

**Actual behavior:**

It alternates between `string` and `number` and then breaks after while.
From that point, it just sticks with `string`.

No errors given.

**Playground Link:**

[Playground](https://www.typescriptlang.org/play/#code/C4TwDgpgBAIgTgewM7AgHgCpQgD1QOwBMkoBvHALnwFcBbAIwjgB8U4BLfAcwF8A+KAF4yAKCjioANwCGAG2oQKUDABoxEgPQAqLeolaoAZQAWCarMJQ5qOPmmoojYAHcIEfFDacuVolBoMTHriWhrBhIgoirCRqGjkFAAUGADaAEQ4aQC62HjuxP50jHBQAPyewBzcUEoBxQCU-CI8ANwiIoQQAMay0nDQXQj4KFARyKhK8OPoCXVM-G0iGhpzcCKDw8BQAPoADEKjsRAAdGNRp0cX01fnZ6g393cnTw-PR8Gvny-flz-Xf7d3hIoF8jm0NiNtgBGA57V4tKDLLzcdZDSEAJlhUPhiJWRSCEK22wAzLD0TikZVvO1CTsACyw4kUjQAURwkC6qEsqxUVk51DkFSqXF5pTFqM2OwArLC6czkVwJZCAGywqXyqko2nbADssOVGuFSqJAA5YTrDdTtQBOWEmy0oynCqAAMwQ-UkTF8lggnrgxygAHVjExoOwSAADVYR0pAA)

**Related Issues:** 

It is similar to https://github.com/microsoft/TypeScript/issues/32707
Except, it gives up and resolves the type to `string`, rather than `any`.

**Comment by AnyhowStep** [comment]
I know, I know, recursive type alias, bad.

Interfaces break, too,
```ts
interface Droste<T extends {x:number|string}> {
    value: T,
    /**
     * Should alternate between string and number
     */
    droste: Droste<{x:(T["x"] extends number ? string : number)}>
};

declare const droste: Droste<{x:number}>;

//number
const _0 = droste.droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste;
const _1 = _0.droste; //string
const _2 = _1.droste; //number
const _3 = _2.droste; //string

const _4 = _3.droste; //number
const _5 = _4.droste; //Expected string, actual number, ???
const _6 = _5.droste; //number
const _7 = _6.droste; //number
const _8 = _7.droste; //number
const _9 = _8.droste; //number
//number forever and ever. Where is `string`?
```
[Playground](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgCJQPYGdIB4AqyEAHpCACZbIDexAXCAK4C2ARtAD45SgDmAvgD4aAKGTjkANzgAbRhDrJ8AGjESA9ACpNaiZuQBlABYZGM8slmQoIOJGTswAdwgQQybn0sVkTNtF1xTXVA8kwcBTRwvFo6AAp8AG0AImJkgF0iUjdKXxZ2KGQAfg8wHhBeZEU-AoBKIRF+AG4REXIIBBk4KBQEDBAcZDDsSEV0EYhcWJroIRaRdXUZqBE+gbBkAH0ABmQAXiHoiAA6YYjTo4uJq-OzyBv7u5Onh+ejwNfPl+-Ln+u-27vCTIL5HFprQabACM+y221eTWQi08FVW-UhACZYdCEUilvkAhCNpsAMzYjG45FlPitIlbAAs2JJlPx-hWdM2AFZsfSWQBRYgABw6kAsKN4yksCDAjFkeTZkqKSrR6y2ADZsZyWcsVZCAOzYtXagns9HEgAc2L1xrZuuJAE5seabQUFqyCsgYBgepJoN4LBBfVBjsgAOpGaAoYBUAAG4pjRSAA)

**Comment by webstrand** [comment]
Looks like it's falling afoul of the instantiation limit, but somehow silently failing.

```ts
type check<T> = T extends Droste<{ x: string }> ? true : false;
type a = check<typeof _1>; // true
type b = check<typeof _2>; // false 
type c = check<typeof _3>; // Type instantiation is excessively deep and possibly infinite.
```

**Comment by AnyhowStep** [comment]
It seems like the workaround outlined in https://github.com/microsoft/TypeScript/issues/32707#issuecomment-518347966 works,
```ts
/**
 * Uses the workaround discovered here,
 * https://github.com/microsoft/TypeScript/issues/32707#issuecomment-518347966
 * 
 * Now, everything resolves correctly.
 */
type NextDrosteFix<X extends number|string> = (
    Droste<{x:(X extends number ? string : number)}>
);
type NextDroste<T extends {x:number|string}> = (
    NextDrosteFix<T["x"]>
);
interface Droste<T extends {x:number|string}> {
    value: T,
    /**
     * Should alternate between string and number
     */
    droste: NextDroste<T>
};

declare const droste: Droste<{x:number}>;

//number
const _0 = droste.droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste.droste.droste.droste.droste.droste.droste.droste.droste
    .droste.droste;
const _1 = _0.droste; //string
const _2 = _1.droste; //number
const _3 = _2.droste; //string

const _4 = _3.droste; //number
const _5 = _4.droste; //string
const _6 = _5.droste; //number
const _7 = _6.droste; //string
const _8 = _7.droste; //number
const _9 = _8.droste; //string

```
[Playground](https://www.typescriptlang.org/play/#code/PQKhCgAIUhVBnApvSAXAFoyB3A9gJwGsBDfXAVwDsATSagS3gGNcA3RfRWzTgGihjpUqAA7wAXMGABzehnIAjAHQsAtsFX0mZeLgBmqYABUAniMQBlbfRGHG8csmABmAEwB2AAzuAxPcdqqoiUqAC0AKwAjAAczgAs7gCcAGzJApDpAHK42LyQiOz4Jhj0lNKQnLoANuwoLPicTKhVJkoCwOCoZliZiAAeqAAiOqiIAGL0fQA8ABr5A8HUKJTkqgocAD7wqPil0gB8kAC8kAAUUJCXw7jbiFMA3n3ip3P9ozTLq+v4kAD8kNtdmVIOJICs1hwAJQAX324EhAG5Ot1IL0BtdblMjPN3ktII9xODvlsdntYcczhdLmihiNxpMsQBtABEfWZAF04YjwKVRvg9MQmFgMaMsTjFigCUTNoCyYd7lTIKxiFVHKCjPxLpdQBAtZcYBZ0BQqrQVXzKMRRpB1qhsIhggDScDiDQwV8OIroB09dQ6aCaSK7kY4dCkeBqIgmFVSFgWJRtnQ-ZBAw8ntL8LCw1J0+A4wmAPqeCm+m6jJQl27lulV0uIGuVitlxt15v1pvV5uKtstju92vdgetod925d4e1pF51CQfORCmF7sIyBSWVlXO4ePT-OueeRRfL4A5qcz5zz1z7ldO6TgdebmdxefOC+H934W8F8LzuLP1fX4-55J53CZ8jw3At3HnZIfyvd8t2ied3BA19YJnRJ52iaCgWvIA)

And, as always, I know it works but I don't know **why** it works.

**Comment by jack-williams** [comment]
If I had to guess (and it really is a guess), there is a caching of instantiations done on type id. With your original example you create a new object literal type (with a new id) for each recursive instantiation that will have no cached value.

In the modified example each the recursion goes through `NextDrosteFix` which is instantiated with a primitive type (and primitive types are interned). `NextDrosteFix` will be cached for the instantiations at `string` and `number`.

**Comment by AnyhowStep** [comment]
Are array types also considered primitive types for interning purposes?

Array types never seem to give problems. Only object literal types.

**Comment by jack-williams** [comment]
I think arrays and tuples are cached, but I would wait on confirmation.

**Comment by RyanCavanaugh** [comment]
![Truman hits the edge](https://i.giphy.com/media/HP1FdOEalARC8/giphy.webp)

**Comment by RyanCavanaugh** [comment]
Fixed by #45025.

With the added check that `_3.droste` is assignable to `Droste<{ x: number }>`, `4.5.0-dev.20210817` reports TS2322 and TS2589 ("Type instantiation is excessively deep and possibly infinite"). `4.5.0-dev.20210818` is clean, and the current nightly (`7.1.0-dev.20260902.1`) is also clean. The increased instantiation-depth limit allows the chained `droste` accesses to continue alternating as expected.
